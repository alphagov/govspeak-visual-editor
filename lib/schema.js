import { Schema } from "prosemirror-model";
import nodeDefinitions from "./nodes";
import markDefinitions from "./marks";

export default function constructSchema(options) {
  const nodes = Object.fromEntries(
    nodeDefinitions.map((node) => [node.name, node.schema(options)]),
  );

  const marks = Object.fromEntries(
    markDefinitions.map((mark) => [mark.name, mark.schema(options)]),
  );

  return new Schema({
    nodes,
    marks,
  });
}
