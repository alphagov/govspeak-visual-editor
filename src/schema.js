import { addListNodes } from "prosemirror-schema-list"
import { Schema } from "prosemirror-model"
import { schema as basicSchema } from "prosemirror-schema-basic"

// Mix the nodes from prosemirror-schema-list into the basic schema to
// create a schema with list support.
export default new Schema({
  nodes: addListNodes(basicSchema.spec.nodes, "paragraph block*", "block"),
  marks: basicSchema.spec.marks,
});
