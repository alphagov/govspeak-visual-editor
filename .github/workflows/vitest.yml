name: Vitest
on: [pull_request]
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-node@v3
        with:
          node-version: 18
      - name: Install dependencies
        run: npm ci
      - name: Run Unit tests
        run: npm run test
