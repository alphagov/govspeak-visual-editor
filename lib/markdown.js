import { MarkdownSerializer } from "prosemirror-markdown";
import customNodes from "./nodes.js";

const defaultNodesSerializer = {
  blockquote(state, node) {
    state.wrapBlock("> ", null, node, () => state.renderContent(node));
  },
  code_block(state, node) {
    // Make sure the front matter fences are longer than any dash sequence within it
    const backticks = node.textContent.match(/`{3,}/gm);
    const fence = backticks ? backticks.sort().slice(-1)[0] + "`" : "```";

    state.write(fence + (node.attrs.params || "") + "\n");
    state.text(node.textContent, false);
    // Add a newline to the current content before adding closing marker
    state.write("\n");
    state.write(fence);
    state.closeBlock(node);
  },
  heading(state, node) {
    state.write(state.repeat("#", node.attrs.level) + " ");
    state.renderInline(node, false);
    state.closeBlock(node);
  },
  horizontal_rule(state, node) {
    state.write(node.attrs.markup || "---");
    state.closeBlock(node);
  },
  bullet_list(state, node) {
    state.renderList(node, "  ", () => (node.attrs.bullet || "*") + " ");
  },
  ordered_list(state, node) {
    const start = node.attrs.order || 1;
    const maxW = String(start + node.childCount - 1).length;
    const space = state.repeat(" ", maxW + 2);
    state.renderList(node, space, (i) => {
      const nStr = String(start + i);
      return state.repeat(" ", maxW - nStr.length) + nStr + ". ";
    });
  },
  list_item(state, node) {
    state.renderContent(node);
  },
  paragraph(state, node) {
    state.renderInline(node);
    state.closeBlock(node);
  },
  image(state, node) {
    state.write(
      "![" +
        state.esc(node.attrs.alt || "") +
        "](" +
        node.attrs.src.replace(/[()]/g, "\\$&") +
        (node.attrs.title
          ? ' "' + node.attrs.title.replace(/"/g, '\\"') + '"'
          : "") +
        ")",
    );
  },
  hard_break(state, node, parent, index) {
    for (let i = index + 1; i < parent.childCount; i++)
      if (parent.child(i).type !== node.type) {
        state.write("\\\n");
        return;
      }
  },
  text(state, node) {
    state.text(node.text, !state.inAutolink);
  },
};

const customNodesSerializer = Object.fromEntries(
  customNodes.map((node) => [node.name, node.toGovspeak]),
);

const nodesSerializer = Object.assign(
  defaultNodesSerializer,
  customNodesSerializer,
);

export const markdownSerializer = new MarkdownSerializer(nodesSerializer, {
  em: { open: "*", close: "*", mixable: true, expelEnclosingWhitespace: true },
  strong: {
    open: "**",
    close: "**",
    mixable: true,
    expelEnclosingWhitespace: true,
  },
  link: {
    open(state, mark, parent, index) {
      state.inAutolink = isPlainURL(mark, parent, index);
      return state.inAutolink ? "<" : "[";
    },
    close(state, mark, parent, index) {
      const { inAutolink } = state;
      state.inAutolink = undefined;
      return inAutolink
        ? ">"
        : "](" +
            mark.attrs.href.replace(/[()"]/g, "\\$&") +
            (mark.attrs.title
              ? ` "${mark.attrs.title.replace(/"/g, '\\"')}"`
              : "") +
            ")";
    },
    mixable: true,
  },
  code: {
    open(_state, _mark, parent, index) {
      return backticksFor(parent.child(index), -1);
    },
    close(_state, _mark, parent, index) {
      return backticksFor(parent.child(index - 1), 1);
    },
    escape: false,
  },
});

function backticksFor(node, side) {
  const ticks = /`+/g;
  let m;
  let len = 0;
  if (node.isText)
    while (m === ticks.exec(node.text)) len = Math.max(len, m[0].length);
  let result = len > 0 && side > 0 ? " `" : "`";
  for (let i = 0; i < len; i++) result += "`";
  if (len > 0 && side < 0) result += " ";
  return result;
}

function isPlainURL(link, parent, index) {
  if (link.attrs.title || !/^\w+:/.test(link.attrs.href)) return false;
  const content = parent.child(index);
  if (
    !content.isText ||
    content.text !== link.attrs.href ||
    content.marks[content.marks.length - 1] !== link
  )
    return false;
  return (
    index === parent.childCount - 1 ||
    !link.isInSet(parent.child(index + 1).marks)
  );
}
