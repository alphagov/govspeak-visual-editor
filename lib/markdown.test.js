import { expect, test } from "vitest";
import { doc, p, builders } from "prosemirror-test-builder";
import { markdownSerializer } from "./markdown";
import schema from "../lib/schema.js";

const {
  address,
  call_to_action,
  contact,
  example_callout,
  information_callout,
  warning_callout,
} = builders(schema);

test("Custom nodes", () => {
  const state = doc(
    address(p("address")),
    call_to_action(p("call to action")),
    contact(p("contact")),
    example_callout("example callout"),
    information_callout("information callout"),
    warning_callout("warning callout"),
  );

  expect(markdownSerializer.serialize(state)).toMatchSnapshot();
});
