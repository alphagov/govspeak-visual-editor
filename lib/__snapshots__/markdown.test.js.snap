// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Custom nodes 1`] = `
"$A

address

$A

$CTA

call to action

$CTA

$C

contact

$C

$E

example callout$E

^information callout^

%warning callout%"
`;

exports[`Links and email links 1`] = `
"[link](example.com)

<contact@example.com>

[link](mailto:contact@example.com)"
`;
